### ISSUE TITLE
Solution slides / editorials

### ISSUE THREAD

**Issue by niemela** [Kattis/problem-package-format#70]
We should define a place to store solution slides / editorials.

At a minimum in the form of a PDF, but preferably (?) as LaTeX and/or Markdown (i.e. similar to problem statements). This of course introduces more of the #31 issues.

**Comment by evouga** [comment]
Yes, if problem packages are to be used for long-term archiving then I agree it's important to preserve the judges' editorials. 

Do we want to insist on some kind of Beamer template? Otherwise I don't think per-problem .tex really makes sense. Maybe an "editorial" folder that can contain anything (including source) and is preserved verbatim.

**Comment by eldering** [comment]
Doesn't this make more sense at the contest level and this in the Contest Package https://ccs-specs.icpc.io/draft/contest_package ?

**Comment by niemela** [comment]
@RagnarGrootKoerkamp Could you describe what you're currently doing, that sounded like a good approach.

**Comment by niemela** [comment]
> Doesn't this make more sense at the contest level and thus in the Contest Package

I don't think so. The solution to the problem is useful wherever you have the problem, and is clearly related to the problem.

**Comment by RagnarGrootKoerkamp** [comment]
BAPCtools supports `problem_statement/solution.en.tex`, and adds `contest/solution_{header,footer}.tex` files that are included before the first problem and after the last problem.

I like having them as separate files per problem. That also allows automatic reordering by difficulty at the end of a contest using the API.

I agree that `problem_statement` may not be the most intuitive name for the directory it's in, but it's nice to have it next to the statement I think.

**Comment by jsannemo** [comment]
I like that.

**Comment by jsannemo** [comment]
So I think there's a general preference towards having a partial tex/md/whatever file (or compiled PDF) in the same way we handle statements?

There are some options on exactly where it should be:
- problem_statement/solution.en.FORMAT
- {tutorials,solution,spoiler}/{the same thing}.en.FORMAT

Aside from me and Ragnar, are there any preferences/thoughts on that? @niemela @evouga @Tagl (random subset of people).

**Comment by RagnarGrootKoerkamp** [comment]
I'm ok with either reusing or making a new directory. If we making a new one, I strongly prefer `solution/solution.en.tex`.

Note that specifying exactly what commands are allowed in solution slides is much more tricky than for statements. We do a lot of customization there to show statistics coming from the API (graph of solves over time, number of submissions/AC before freeze, number of pending submissions, shortest submitted solution, ...).

**Comment by jsannemo** [comment]
It's the same as with problem.tex though: the subset is only a promise that a "more" best-effort can be made, not that it's required. For my judge I plan to warn if going outside my supported subset, but compile it to a pdf and offer it for download.

**Comment by niemela** [comment]
> Note that specifying exactly what commands are allowed in solution slides is much more tricky than for statements. We do a lot of customization there to show statistics coming from the API (graph of solves over time, number of submissions/AC before freeze, number of pending submissions, shortest submitted solution, ...).

Would the API statistics ones mainly be in the "core" LaTeX though?

In the cases where they are, would it be possible to get some "ignore if note available" semantics to work?

Could you share a few examples of how you currently do this?

**Comment by RagnarGrootKoerkamp** [comment]
```
\begin{frame}
    \frametitle{\problemtitle}

    \begin{block}{Problem}
       ...
    \end{block}
    \begin{block}{Pitfalls}
      ...
    \end{block}
    \solvestats
\end{frame}
```

The submissions over time graph is included in the `\frametitle` so that works OK, but we do have the `\solvestats` at the end of the slide that cannot really be automated, since we only want to show that at the bottom of the last slide for each problem typically. But defaulting it to empty works fine indeed.

Another thing I just noticed is that last nwerc we also made a `problem_statement/problem-slide.tex` that was compiled into a separate presentation of brief problem statements that was used in the livestream and such.

**Comment by niemela** [comment]
> Another thing I just noticed is that last nwerc we also made a `problem_statement/problem-slide.tex` that was compiled into a separate presentation of brief problem statements that was used in the livestream and such.

Interesting. What did that look like?

**Comment by RagnarGrootKoerkamp** [comment]
What you see here in the bottom-left:
https://www.youtube.com/live/aDsW8J8P7go?feature=share&t=14270

**Comment by mpsijm** [comment]
Slightly late to the party, it seems :stuck_out_tongue: The problem slides for NWERC can be found at this (hidden) link: https://chipcie.wisv.ch/archive/2023/nwerc/problem-slides.pdf (also for [2022](https://chipcie.wisv.ch/archive/2022/nwerc/problem-slides.pdf)). My branch for compiling them can be found at https://github.com/mpsijm/BAPCtools/tree/problem-slides (I update this branch roughly once a year, namely around my "ICPC contest season" in September–November, but let me know if you want to try it out on the latest version and I'll rebase :slightly_smiling_face:)

I'm slightly in favour of keeping all TeX files (for problem statement, solution, and problem slides) together in `problem_statement`, for the benefit of sharing images and such, as mentioned by @Tagl in https://github.com/Kattis/problem-package-format/issues/70#issuecomment-1654526617.

However, if we do end up splitting directories, `solution` sounds to me like a good name to store the `solution.en.tex` :slightly_smiling_face: We'll just have to do some symlinking to share assets between the problem statement and the solution slides :stuck_out_tongue: Also in this case, I'll have to figure out what to do with the problem slide. Maybe I'll just keep them in `problem_statement`? Not sure whether this should be part of the spec, or that this can be a BAPCtools-only thing.

**Comment by mzuenni** [comment]
I would also be in ravor of keeping them together and share resources.

**Comment by niemela** [comment]
> Also in this case, I'll have to figure out what to do with the problem slide.

What is the "problem slide". Is that not the same as `solution.*.tex`?

**Comment by RagnarGrootKoerkamp** [comment]
At NWERC we've been using:
- the problem statement for contestants: `problem_statement/problem.en.tex`
- the solution: `problem_statement/solution.en.tex`
- a 'problem slide', succinctly explaining the problem for use in the livestream: `problem_statement/problem-slide.tex`.

I'd also slightly prefer to keep everything together. What are some reasons to prefer separating them?

**Comment by niemela** [comment]
> I'd also slightly prefer to keep everything together. What are some reasons to prefer separating them?

I can only think of two reasons (neither of them *very* good):
- The contents of `problem_statement/` is meant for the contestants, i.e. it's "public" information after the contest has started. The `solution/` very much isn't. There are situations where you might want to share the problem statement with people that you do not want to share the solution with, e.g. translations in an IOI-like situation.
- We tend to put separate things in separate (top level) directories. There is included files iff `include/` exists, there are attachements iff `attachments/ exists, there are solution descriptions iff `solution/`exists.

**Comment by mzuenni** [comment]
Why did we end up with putting things in separate directories here? As far as I can see, most people were in favor of making it the same directory.

Regarding the two not very good counterarguments:
> The contents of problem_statement/ is meant for the contestants, i.e. it's "public" information after the contest has started. The solution/ very much isn't. There are situations where you might want to share the problem statement with people that you do not want to share the solution with, e.g. translations in an IOI-like situation.

This doesn't need to be decided on a directory level, i.e., we can say the same about files: `probelm_statement.<?>` are public files and `solution.<?>` are not.

> We tend to put separate things in separate (top level) directories. There is included files iff include/ exists, there are attachements iff attachments/ exists, there are solution descriptions iff solution/exists.

Yes, but those things are also completely independent. For statements and solution, this is pretty much not the case most of the time, i.e., at NWERC and GCPC we almost always reuse illustrations and graphics from the statement for the solutions. Sure, the user could symlink those files if he wants to reuse them, but it is much more convenient to just have access to the same files.

I would propose to move this into one directory again as there don't seem to be many counterarguments and most people seemed to be in favor of this. I will write a PR if people agree.

**Comment by niemela** [comment]
I'm not at all sure about this.

Shared resources is a good argument for.

As I said above, I think it's problematic that `problem_statement` contains both very public and very secret data.

> This doesn't need to be decided on a directory level, 

Agreed.

> i.e., we can say the same about files: `probelm_statement.<?>` are public files and `solution.<?>` are not.

But that won't work.  All files other than `problem.*.tex` and `solution.*.tex` must be assumed to be needed by both, so it makes it impossible to safely share the problem statement for e.g. translation.

**Comment by evouga** [comment]
I don't have very strong feelings here. I _slightly_ prefer separate directories, for both the safety reasons Fredrik mentions and just to keep a clear delineation of what material goes in which folder (not currently a potential source of confusion, but might become one if in the future people start putting other problem-statement-like material in `problem_statement` for things like dark mode, screen readers, etc.

But I really don't care much either way.

**Comment by niemela** [comment]
Yeah, the more I think about this the more I'm against merging the directories.

I think it's very good if "separate features" (which admittedly is not a perfectly well defined concept) are being kept as separate as possible, both so they can be safely ignored as needed, and so that it is easy to detect when they are being used.

**Comment by Tagl** [comment]
I prefer separating as well. Symlinks can take care of reuse from my
perspective.

**Comment by niemela** [comment]
> The point here is that from a user point of view they are very much not separate.

I mean from a problem setter (which is definitely one important type of user) they are of course strongly related, but they are clearly separated. You definitely know which is which. But know I'm just nitpicking words, I understand what you're saying.

> Typically, they are both latex file and It's very typical to share resources in latex (I know, we already define so many things which are considered bad practice in the latex world like defining that the document class is magically added and not using the standalone package for this... but fixing those things is a bit late now)

Ok, huge sidetrack here... I don't agree with "we already define so many things which are considered bad practice in the latex world", mainly because I don't agree that we are "defining that the document class is magically added and not using the standalone package for". What we are doing is that we're saying that the `problem.en.tex` file is included in some larger LaTeX document that created the problem set document. This is severely underdefined in the spec, and if someone would be able to define it properly (I think the right approach would be along the lines of `problem.en.tex` has the meaning as if included into \<this given file>, and must work when included into \<this given file> (but you are of course allowed to include it into other things)). This is very reasonable to do in the LaTeX world (IMO).

> > Symlinks can take care of reuse from my perspective.
> 
> Sure they are sufficient but very annoying to use especially since this is not a rare usecase.

Yeah, I totally see that.

> If there is a strong opinion against this, can we at least define that the content of `statement` can be considerd as part of the working directory of the solution?
> To be more specific, say that these file `statement/bob.pdf` exists and `solution/bob.pdf` not. Then it is ok to include `bob.pdf` in the solution.

That's a really good idea! (again IMO). Could we make some statement on conflicting files names (i.e. if we have different `bob.pdf` in both `/statement` and `/solution`), in such a way that it would be a legal implementation for a system to copy `/solution` into (the same directory) as `/statement`?

**Comment by mzuenni** [comment]
> Could we make some statement on conflicting files names (i.e. if we have different bob.pdf in both /statement and /solution), in such a way that it would be a legal implementation for a system to copy /solution into (the same directory) as /statement?

No opinion on the exact wording here. If there are conflicting files I would say that `bob.pdf` refers to the one in the same directory, but we could also forbid this (but I don't see the need for that).
From the latex point of view, I would want to say that when compiling solutions the environment variable `TEXINPUTS` contains the statement folder.

**Comment by evouga** [comment]
This sounds pretty clean: we set `TEXINPUTS` to include `problem_statement`, but don't actually copy any files around?

**Comment by mzuenni** [comment]
>This sounds pretty clean: we set `TEXINPUTS` to include `problem_statement`, but don't actually copy any files around?

I think these are implementation details which we should not mention. What we define should also work for markdown.

So we should say something like:
the working directory contains every file from `statement` which is not also present in `solution`.

**Comment by mzuenni** [comment]
It might even be cleaner to just say that `\graphicspath` always contains the statement directory, i.e., specifically allow to include graphics

**Comment by RagnarGrootKoerkamp** [comment]
I'd avoid `\graphicspath`. It doesn't work for all packages / all types of includes. (IIRC either `\includesvg` or some variant of `\lstinputlisting` didn't pick it up anyway.)

**Comment by mzuenni** [comment]
> I'd avoid \graphicspath. It doesn't work for all packages / all types of includes. (IIRC either \includesvg or some variant of \lstinputlisting didn't pick it up anyway.)

It is only supposed to work for graphics. `\input@path` should work for most other things.
(The current BAPCtools implementation copies the main `.tex` file into a `tmp` directory and only supports graphics includes from `problem_statement` by setting `\graphicspath`)

The issue with `TEXINPUTS` is that it will break as soon as you intend to build a single pdf for multiple problems, e.g., a contest.pdf.

**Comment by RagnarGrootKoerkamp** [comment]
> The issue with `TEXINPUTS` is that it will break as soon as you intend to build a single pdf for multiple problems, e.g., a contest.pdf.


Ah right; so then we must choose between copying everything, or speccing `\graphicspath`. I'm neutral on this, but I guess graphicspath is cleaner in the implementation.

**Comment by mzuenni** [comment]
Two more questions:

1. should it be `statement/statement.tex` for consistency?
2. should it be `statements/` and `solutions` since we support multiple languages?

**Comment by thorehusfeldt** [comment]
1. I am _in favour_ of `statement/statement.{en, da}.tex` (rather than `problem_statement/problem.en.tex`)
2. I _mildly against_ `statements/statement.en.tex` with the plural in the directory name. (I do find the plural slightly silly, and it not clear to me that translations are _different_ statements. After all, _Krieg und Frieden_ and _Война и миръ_ are the same book, not two books.)
3. I am _neutral_ toward `solution/solution.en.tex` versus `statement/solution.en.tex` (with separate directories). Solution slides are _nice to have_ (unlike, say, `data/sample` or `output_validator`, whose absence makes the problem unusable) and may not deserve top level presence in the problem package format.

**Comment by niemela** [comment]
I consider this closed by #374.

As always, feel free to reopen if you disagree.
